### DISCUSSION TITLE
Making post requests to API

### DISCUSSION THREAD

**Discussion by tGitm** [nextgenhealthcare/connect#5011] [Q&A] [answered]
Hi 👋 

I would like to make https requests on API in mirth channel. Username, password and .pfx certificate with password are required for successful authentication. As response I would get HL7 message which I have to send to another channel for reading it's data (channel already working). Can I do something like that in Mirth? 

This is code from C# sample app from provider.
![image](https://user-images.githubusercontent.com/47832737/153146391-25cc5166-ddce-4ae0-8214-d5637ebd3d54.png)

Thanks for help!

**Comment by jonbartels** [comment]
If the certificate is being used for mutual TLS please review the options for SSL in MC at: https://gist.github.com/jonbartels/8abd121901eb930f46245d9ef0f5710e . You will either need a plugin or an SSL tunnel or an SSL proxy I think.

I hit `https://azer.winter.weda.fr/` and it does seem to want mutual TLS. It is possible there is some other authentication scheme for this API that uses a keypair from a PFX but does not use TLS, you would have to read or share the API documentation to know.

**Reply by tGitm** [reply]
Thanks for answer. Here is all I got of documentation:
[documentation_WEDA.pdf](https://github.com/nextgenhealthcare/connect/files/8038742/documentation_WEDA.pdf)

**Reply by jonbartels** [reply]
That documentation helps.

> Protocol: Secure HTTPs with TLS 1.2

This means you need TLS. The gist I posted explains how to do that. If one option sounds better for you, the community here can help advise on a specific option. I like Zens SSL plugin for this, https://consultzen.com/zen-ssl-extension/ . Zen is my employer so I am biased.

> Headers

The header calculation is pretty straightforward. I would write a code template to compute that in MC. Then call that header in a deploy script for the channel and put it into the `globalChannelMap` for your channel.

Then when you build your HTTP connection just set the Header field in the configuration for that connection. 

> Body

This looks like just HL7 in a POST body. Make sure you're building correct HL7 messages and this is easy, routine work for Mirth Connect.

**Reply by tGitm** [reply]
I will try with this plugin: https://github.com/tonygermano/connect-plugins/tree/master/basicssl

> This are instructions:


In order to build, you must place these mirth jars in a lib subdir, mirth-server.jar, donkey-server.jar, http-server.jar, and http-shared.jar from your mirth release are needed for this project. The mirth and donkey files are located in server-lib, and the http files are in extensions/http in your mirth install directory. See build.gradle for more details.

In order to build, you must place these mirth jars in a lib subdir, mirth-server.jar, donkey-server.jar, http-server.jar, and http-shared.jar from your mirth release are needed for this project. The mirth and donkey files are located in server-lib, and the http files are in extensions/http in your mirth install directory. See build.gradle for more details.


> I put my files in lib subdir in basicssl folder and run ./gradlew build in root folder of plugin

![image](https://user-images.githubusercontent.com/47832737/153820351-650cbe09-57ad-4c95-9fa8-6b1a14469ba1.png)


> After running gradlew build I'm facing an error:

![image](https://user-images.githubusercontent.com/47832737/153568899-84e88a63-bf2a-4d29-a6a9-ad7523a46b9c.png)

`
FAILURE: Build failed with an exception.
`
* What went wrong:
Could not initialize class org.codehaus.groovy.reflection.ReflectionCache

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

**So I changed gradlew-wrapper.properties to run gradle 6.3 and it works, but now Iam facing another error with kotlin:**

`
Task :server:compileKotlin FAILED
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (21, 1): Class 'BasicSslPlugin' is not abstract and does not implement abstract member public abstract fun getObjectsForSwaggerExamples(): (Mutable)Map<String!, Any!>! defined in com.mirth.connect.plugins.ServicePlugin
FAILURE: Build failed with an exception.
`
`
What went wrong:
Execution failed for task ':server:compileKotlin'.
Compilation error. See log for more details
`


Thanks again! 🤘

**Reply by jonbartels** [reply]
@tonygermano can you help out @tGitm with a build error for your ssl extension?

**Reply by tGitm** [reply]
@tonygermano sorry for another ping, but I can't get it to work :)

**Reply by jonbartels** [reply]
Oh! You're running into https://github.com/nextgenhealthcare/connect/pull/4868

**Reply by jonbartels** [reply]
Alter this file https://github.com/tonygermano/connect-plugins/blob/master/basicssl/server/src/main/kotlin/io/github/tonygermano/connect/plugins/basicssl/BasicSslPlugin.kt to add

```
override fun getObjectsForSwaggerExamples() {
        return null;
    }
```

(Warning - I am not a Kotlin developer)

**Reply by tGitm** [reply]
> Alter this file https://github.com/tonygermano/connect-plugins/blob/master/basicssl/server/src/main/kotlin/io/github/tonygermano/connect/plugins/basicssl/BasicSslPlugin.kt to add
> 
> ```
> override fun getObjectsForSwaggerExamples() {
>         return null;
>     }
> ```
> 
> (Warning - I am not a Kotlin developer)

Now I am facing this error:
`
Task :server:compileKotlin FAILED e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (42, 18): Return type of 'getObjectsForSwaggerExamples' is not a subtype of the return type of the overridden member 'public abstract fun getObjectsForSwaggerExamples(): (Mutable)Map<String!, Any!>! defined in com.mirth.connect.plugins.ServicePlugin' e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (43, 16): Null can not be a value of a non-null type Unit
`

**Reply by jonbartels** [reply]
It looks like its missing a return type.

What you have to do is update Tonys plugin for a newer version of MC which requires the plugin to define `getObjectsForSwaggerExamples()`. It can return null, but the function has to be defined correctly for it to compile and run.

Maybe this? I'm on a different computer and I don't have a clone to test with. 
```
override fun getObjectsForSwaggerExamples() : Map<String, Any> {
        return null;
    }
```

**Reply by tGitm** [reply]
This gives me a lot of new errors

`
Task :server:compileKotlin FAILED
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (39, 12): Unresolved reference: mirth
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (40, 12): Unresolved reference: mirth
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (41, 12): Unresolved reference: mirth
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (42, 12): Unresolved reference: mirth
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (43, 12): Unresolved reference: mirth
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (44, 12): Unresolved reference: mirth
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (45, 12): Unresolved reference: mirth
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (46, 12): Unresolved reference: mirth
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (48, 33): Unresolved reference: HttpConfiguration
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (50, 43): Unresolved reference: ControllerFactory
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (52, 5): 'configureConnectorDeploy' overrides nothing
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (52, 54): Unresolved reference: Connector
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (53, 26): Unresolved reference: HttpDispatcher
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (58, 5): 'configureConnectorUndeploy' overrides nothing
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (58, 56): Unresolved reference: Connector
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (60, 5): 'configureReceiver' overrides nothing
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (60, 47): Unresolved reference: HttpReceiver
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (95, 45): Unresolved reference: MirthSSLUtil
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (96, 48): Unresolved reference: MirthSSLUtil
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (99, 5): 'configureDispatcher' overrides nothing
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (99, 49): Unresolved reference: HttpDispatcher
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (99, 86): Unresolved reference: HttpDispatcherProperties
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (101, 5): 'configureSocketFactoryRegistry' overrides nothing
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (101, 61): Unresolved reference: ConnectorPluginProperties
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (102, 32): Unresolved reference: MirthSSLUtil
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (103, 35): Unresolved reference: MirthSSLUtil
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicHttpsConfiguration.kt: (108, 5): 'getRequestInformation' overrides nothing
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (11, 12): Unresolved reference: mirth
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (12, 12): Unresolved reference: mirth
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (13, 12): Unresolved reference: mirth
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (21, 24): Unresolved reference: ServicePlugin
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (23, 43): Unresolved reference: ControllerFactory
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (25, 5): 'getPluginPointName' overrides nothing
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (27, 5): 'start' overrides nothing
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (29, 5): 'stop' overrides nothing
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (31, 5): 'init' overrides nothing
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (36, 5): 'update' overrides nothing
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (38, 5): 'getDefaultProperties' overrides nothing
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (40, 5): 'getExtensionPermissions' overrides nothing
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (40, 51): Unresolved reference: ExtensionPermission
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (42, 5): 'getObjectsForSwaggerExamples' overrides nothing
e: E:\documents\MESI\addons\connect-plugins\basicssl\server\src\main\kotlin\io\github\tonygermano\connect\plugins\basicssl\BasicSslPlugin.kt: (43, 16): Null can not be a value of a non-null type Map<String, Any>
`

**Comment by pacmano1** [comment, marked as answer]
If you revisit https://github.com/tonygermano/connect-plugins/tree/master/basicssl, you will see the specific reference to http listeners, not http senders so it won't work for you.  Nor does it do mutual TLS if that was needed.   

Perhaps revisit the link @jonbartels posted above with the assorted options for TLS?

**Reply by tGitm** [reply]
Oh.. I missed this information, sorry. Have a nice day :)
